upstairs/src/impacted_blocks.rs

```diff
@@ -2,96 +2,237 @@
+        })
     }
+}
 
-    /// Return true if this list of impacted blocks overlaps with another.
-    pub fn conflicts(&self, other: &ImpactedBlocks) -> bool {
-        for shared_key in self
-            .blocks
-            .keys()
-            .filter(|key| other.blocks.contains_key(key))
-        {
-            // TODO any way to avoid this clone?
-            let bv = self.blocks[shared_key].clone();
-            if bv.bitand(&other.blocks[shared_key]).any() {
-                return true;
+impl ExactSizeIterator for ImpactedBlockIter {
+    fn len(&self) -> usize {
+        match self.active_range {
+            None => 0,
+            Some((fst, lst)) => {
+                let extents = lst.extent_id - fst.extent_id;
+                (extents * self.extent_size + lst.block - fst.block + 1)
+                    as usize
             }
         }
+    }
+}
+
+impl DoubleEndedIterator for ImpactedBlockIter {
+    #[inline]
+    fn next_back(&mut self) -> Option<Self::Item> {
+        match &mut self.active_range {
+            None => None,
+            Some((first, last)) => {
+                let result =
+                    (last.extent_id, Block::new(last.block, self.block_shift));
 
-        false
+                // If first == last, then the iterator is now done. Otherwise,
+                // increment first.
+                if first == last {
+                    self.active_range = None
+                } else if last.block == 0 {
+                    last.block = self.extent_size - 1;
+                    last.extent_id -= 1;
+                } else {
+                    last.block -= 1;
+                }
+
+                Some(result)
+            }
+        }
     }
+}
+
+impl FusedIterator for ImpactedBlockIter {}
 
-    /// Return a list of impacted extents
-    #[cfg(test)]
-    pub fn extents(&self) -> Vec<&u64> {
-        self.blocks.keys().collect()
+impl ImpactedBlocks {
+    /// Create a new ImpactedBlocks range from first and last impacted blocks
+    /// (inclusive) Returns Empty if last_impacted < first_impacted.
+    pub fn new(
+        first_impacted: ImpactedAddr,
+        last_impacted: ImpactedAddr,
+    ) -> Self {
+        if first_impacted <= last_impacted {
+            ImpactedBlocks::InclusiveRange(first_impacted, last_impacted)
+        } else {
+            ImpactedBlocks::Empty
+        }
     }
 
-    /// Returns (extent id, block) tuples
-    pub fn tuples(&self) -> Vec<(u64, Block)> {
-        let mut result = Vec::with_capacity(self.len());
+    /// Create a new ImpactedBlocks range starting at a given offset, and
+    /// stretching n_blocks further into the extent. Panics an error if
+    /// `extent_size` is 0. Returns ImpactedBlocks::Empty if n_blocks is 0.
+    pub fn from_offset(
+        extent_size: u64,
+        first_impacted: ImpactedAddr,
+        n_blocks: u64,
+    ) -> Self {
+        debug_assert!(extent_size > 0);
```

Comment:
I was thinking of actually making it be a Result but I could change it back to an assert. I don't really like release-mode assertions because it brings back the ol java fun times of random functions throwing undeclared exceptions, instead of the code actually knowing where all the errors come from and pushing them all through the correct error handling facilities (if only to log them properly before suddenly dying)